### PR TITLE
[Enhancement] Support for more than one resource manager

### LINKED ISSUE #832: [Enhancement] Localization manager with more than one resource manager
## Summary

Currently the localization manager (singleton) can be initialized only with one resource manager. This is often not enough.

We ceurrently use own implementation for this purpose, that accept a list of resource managers (form common library, different app modules).

## API Changes

Non breaking changes:

https://github.com/xamarin/XamarinCommunityToolkit/blob/f62c078146a05ef92e548f71d6cbb22cea89ff8c/src/CommunityToolkit/Xamarin.CommunityToolkit/Helpers/LocalizationResourceManager.shared.cs#L16

```csharp
ResourceManager[] resourceManagers = new ResourceManager[0];
```

https://github.com/xamarin/XamarinCommunityToolkit/blob/f62c078146a05ef92e548f71d6cbb22cea89ff8c/src/CommunityToolkit/Xamarin.CommunityToolkit/Helpers/LocalizationResourceManager.shared.cs#L25-L32

```csharp
public void Init(params ResourceManager[] resources) =>
    resourceManagers = resources;

public void Init(ResourceManager resource, CultureInfo initialCulture)=>
    Init(initialCulture, resource);

public void Init(CultureInfo initialCulture, params ResourceManager[] resources)
{
    resourceManagers = resources;
    currentCulture = initialCulture;
}
```

https://github.com/xamarin/XamarinCommunityToolkit/blob/f62c078146a05ef92e548f71d6cbb22cea89ff8c/src/CommunityToolkit/Xamarin.CommunityToolkit/Helpers/LocalizationResourceManager.shared.cs#L34-L35

```csharp
public string GetValue(string text)
{
    string translated = null;

    foreach (var rm in resourceManagers)
    {
        translated = rm.GetString(text, CurrentCulture);
        if (translated != null)
        {
            break;
        }
    }

    return translated;
    // Or as LINQ
    // return resourceManagers.Select(s => s.GetString(text, CurrentCulture))
    //    .FirstOrDefault(f => f != null);
}
```

## Intended Use Case

Usage with more than one resource manager.

## Who Will Do The Work?

- [x] I am willing to take this on myself
- [ ] Just putting this out there for someone to pick up

### PR DESCRIPTION
**Pull request by maxkoshevoi**
### Description of Change ###

We use two classes to get localized values at runtime: `TranslateExtesion` and `LocalizedString`.
- `LocalizedString` doesn't store or reference any `ResourceManager`, so it basically already supports more then one.
- `TranslateExtesion` Always uses `ResourceManager` from `LocalizationResourceManager`, so this needs to change.

#### What was done ####

- Added new optional `ResourceManager` parameter to `TranslateExtesion`, so that user could specify `ResourceManager` for each string. If none is specified, `LocalizationResourceManager.DefaultResourceManager` is used.
- Renamed `resourceManager` in `LocalizationResourceManager` to `DefaultResourceManager` and made it readonly and internal (so that `TranslateExtesion` could read it).
- Removed indexer and `GetString` from `LocalizationResourceManager` (as suggested by @brminnick [here](https://github.com/xamarin/XamarinCommunityToolkit/pull/919#issuecomment-780233540)) They were only needed to be used in `TranslateExtesion`, and now it no longer uses them.

#### Alternative implementations ####
- `DefaultResourceManager` can be moved to `TranslateExtesion` as nothing else uses it. But then users will need to initialize `TranslateExtesion` as well as `LRM`

### Bugs Fixed ###

- Fixes #832

### API Changes ###

Added: 
 
- `ResourceManager TranslateExtension.ResourceManager { get; set; } = LocalizationResourceManager.Current.DefaultResourceManager`
- `void LocalizationResourceManager.Init(CultureInfo initialCulture);` (since passing resource manager is no longer mandatory)

Changed:

 - `void LocalizationResourceManager.Init(ResourceManager resource)` => `void LocalizationResourceManager.Init(ResourceManager defaultResourceManager)` (renamed parameter)
 
Removed:

- `string LocalizationResourceManager.GetValue(string text)`
- `string LocalizationResourceManager.this[string text]`
- `LocalizedString(LocalizationResourceManager localizationManager, Func<string> generator = null)` (user cannot create new instances of `LRM`, so this constructor doesn't make sense)

### Behavioral Changes ###

User can now pass any resource manager in `TranslateExstension`

### Breaking Changes ###

- Removed indexer and `GetValue` from `LRM`, but they shouldn't be used anyway. And I can bring them back if needed. Their removal is not critical to this PR.
- Renamed `resource` parameter to `defaultResourceManager`.

### PR Checklist ###

- [x] Has tests (if omitted, state reason in description)
- [x] Has samples (if omitted, state reason in description)
- [x] Rebased on top of main at time of PR
- [x] Changes adhere to coding standard

- [ ] Updated [documentation](https://github.com/MicrosoftDocs/xamarin-communitytoolkit)

### CODE REVIEW

**Comment by maxkoshevoi** [comment]
@pictos @brminnick @jfversluis Could you please take a look at this?

**Comment by maxkoshevoi** [comment]
@brminnick I know, you wanted to remove culture from `LRM` as well, but solution you provided:

```cs
public CultureInfo CurrentCulture
{
	get => CultureInfo.DefaultThreadCurrentUICulture ?? CultureInfo.DefaultThreadCurrentCulture;
	set
	{
		if (CurrentCulture == value)
		{
			return;
		}

		CultureInfo.DefaultThreadCurrentUICulture = value;
		CultureInfo.DefaultThreadCurrentCulture = value;
		OnPropertyChanged(nameof(CurrentCulture));
	}
}
```

Works on Android, but doesn't work on UWP (cannot test on iOS). Looks like resource managers in UWP get their default culture from some other place then `CultureInfo.DefaultThreadCurrentUICulture`

**Comment by pictos** [comment]
@maxkoshevoi right now we are focused in um merge the nullable PR asap. So we will review this one right after the nullable is merged (:

**Comment by maxkoshevoi** [comment]
@pictos Sounds great! I just had some spare time, so implemented this one.

File: samples/XCT.Sample/App.xaml.cs [hunk 1]
```diff
@@ -15,7 +14,6 @@ public App()
 
 			LocalizationResourceManager.Current.PropertyChanged += (sender, e) => AppResources.Culture = LocalizationResourceManager.Current.CurrentCulture;
 			LocalizationResourceManager.Current.Init(AppResources.ResourceManager);
-			LocalizationResourceManager.Current.CurrentCulture = new CultureInfo("en");
```
Comment (maxkoshevoi) [inline samples/XCT.Sample/App.xaml.cs:18]: This was added in nullable PR, and it's not necessary. `CurrentCulture` is nullable.

**Comment by pictos** [comment]
@maxkoshevoi, can you rebase your PR against develop?

**Comment by maxkoshevoi** [comment]
@pictos It's already against develop. Is there something else that needs to be done?

**Comment by pictos** [comment]
@brminnick I remember that you will do some spike on this subject, so can you take a look at this PR?

**Comment by maxkoshevoi** [comment]
@AndreiMisiukevich Could you please assign @brminnick as a reviewer here?

**Comment by AndreiMisiukevich** [comment]
@maxkoshevoi done)

**Comment by brminnick** [comment]
FYI - this introduces a breaking change by modifying this constructor of `LocalizedString`:  `public LocalizedString(LocalizationResourceManager localizationManager, Func<string> generator)`.

This should be merged into a major release, e.g. v2.0.0

**Comment by maxkoshevoi** [comment]
@brminnick It seems like there won't be 2.0.0 considering XCT stops accepting new features in September 2021. Should this PR be merged in 1.3.0 (if approved), transferred to `CommunityToolkit.Maui`, or maybe something else?

**Comment by brminnick** [comment]
Thanks! However, we are no longer adding new features to Xamarin Community Toolkit, focusing on the [.NET MAUI Community Toolkit](https://github.com/communitytoolkit/maui).

I've posted more information about the Future Of Xamarin Community Toolkit here: https://devblogs.microsoft.com/xamarin/the-future-of-xamarin-community-toolkit/?WT.mc_id=mobile-0000-bramin

**Pull request closed without merging**